### PR TITLE
[WIP] Plugins

### LINKED ISSUE #2529: [RFC] Plugin system for custom rules
## Motivation

We currently support custom rules as regexes, but often that's not the right tool for the job. Rules implemented with regexes on SwiftLint are usually the slower ones and also the ones with more false positives. Having first-class support to a plugin system would be essential for empowering users to automatically enforce specific patterns that don't necessarily belong to SwiftLint itself.

## Proposal

This proposal describes a flexible plugin system that doesn't require new rules to be written using Swift. We should provide tools to make it possible though, but that will be left as a future enhancement.

A plugin is an executable that will be run by SwiftLint and should respond to certain commands. Communication will happen via stdin and stdout using JSON unless specified.

A plugin maps 1:1 to a rule, so you can't implement more than one rule in the same plugin.

**Note**: so far this is just a proposal, no concrete implementation has started as far as I know.

### Commands

`plugin` represents the path to a generic custom plugin executable in the commands below.

#### `plugin rule_description`

This will return a JSON containing all the required data to create a `RuleDescription`. Will be run by SwiftLint when parsing the configuration and resolving the rules.

**Output:**

```json
{
    "identifier": "my_cool_rule",
    "name": "Cool Rule",
    "kind": "style",
    "non_triggering_examples": [
        "let x = 10"
    ],
    "triggering_examples": [
        "let x = 20"
    ],
    "min_swift_version": "4.1.0"
}
```

#### `plugin lint`

This will receive a JSON as input and should return another JSON containing the violations.

**Input**:
```json
{
    "structure": ... ,
    "syntax_map": ...,
    "path": "file.swift",
    "contents": "",
    "configuration": {
        "cool_number": 10
    }
}
```

Note: `path` is optional. I have mixed feeling whether `contents` should be populated if `path` is present.

`configuration` contains the raw configuration loaded from the configuration YAML file as described below in "Configuring a plugin".

**Output**:
```json
[
    {
        "severity": "warning",
        "reason": "Failed",
        "location": {
            "line": 1,
            "character: 10
        },
        "byte_offset": 203,
        "character_offset": 300
    }
]
```

Note: `location`, `byte_offset` and `character_offset` are exclusive (only one should be provided). `reason` is optional.

#### Alternatives considered

Another option is to have the `lint` command run only once and keep feeding it JSONs. This would have the benefit of being faster (as fixed costs such as loading dynamic libraries would be paid only once) but adds more complexity.

However, simplifying this might lead to unresolvable performance problems in the future, so I'd love more input on this topic.

### Enabling plugins in the configuration

On `.swiftlint.yml` there'll be a new `plugins` key:

```yml
plugins:
    - path/to/plugin
```

### Configuring a plugin

Configuring a plugin will be done in the same way as a configuration is provided to a rule, by adding a key on `.swiftlint.yml`:

```yml
my_cool_rule:
    cool_number: 10
```

All the values inside the key corresponding to the rule identifier will be provided as part of the JSON input as described in the `plugin lint` section.

### Caching

Caching should work without major changes. The only potential issue is that SwiftLint wouldn't be able to see updates on the binary implementation, forcing the user to use `--no-cache` while implementing the custom rule. This could be mitigated by having a `plugin version` command (described below in "Future Enhancements").

### Analyzer

Implement an `analyze` rule will be unsupported by now as it's still experimental.

### `swiftlint rules`

SwiftLint should list rules implemented in plugins in `swiftlint rules`.

However, it's not clear for me what would be the best option to expose the available configuration options. 

The easiest way would be adding another command in the plugin to return a String (similar to `configurationDescription` in the `Rule` protocol).

### `plugin autocorrect`

Autocorrect support is nice but it'll be left as a future enhancement since it should be an additive change. It should work in a similar way as `plugin lint`.

### `plugin version`

This should output to stdout a string containing the version (e.g. `0.1.0`) and exit the program. This will be called on the start of the lint/autocorrect command on SwiftLint to gather all plugin versions and use that to invalidate the violations cache if needed.

### `syntaxMap` and `structure` only when required

Right now, we'll ask SourceKit for the `structure` and `syntaxMap` for every file that needs to be linted using a plugin rule. However, it's common that a rule only need one or another (or even neither!). We could provide a way for the plugin to explicitly say what piece of information it needs and then SwiftLint would only provide that, speeding up the process.

### Docs generation for plugin rules

`plugin rule_description` can return non-triggering and triggering examples that could be used to generate something like Rules.md for the specific plugin.

### Opt-in rules

As described, all rules provided by plugins are enabled by default. We could add support to opt-in rules in plugins in the future (although I'm not quite sure what would be the real use case for it).

### PR DESCRIPTION
**Pull request by marcelofabri**
Will eventually fix #2529.

Posting it here to gather feedback and highlight some of the changes/decisions.

Missing:
- [ ] Handle nested configurations
- [ ] Add tests for caching
- [ ] Show remote rules in `swiftlint rules` (although I don't think this is *necessarily* a blocker)
- [ ] Allow plugins to expose more than one rule (depends on https://github.com/realm/SwiftLint/pull/2536)
- [ ] Documentation

Will be done later:
- [ ] Create APIs on `SwiftLintFramework` that a plugin can use to register a rule. My idea is to allow users to create a rule using the existing types (`Rule`, `ASTRule`, etc) and just call a register method. Something like this:

```swift

struct MyRule: Rule { 
 // ...
}

let plugin = Plugin()
plugin.register(rules: [MyRule.self])
plugin.run()
```

- [ ] Revise public APIs on `SwiftLintFramework`. In order to be productive, we should consider exposing (a lot) of the convenience functions/types/extensions that are currently `internal`. However, this increases the public API surface. I don't think that too many people use `SwiftLintFramework` directly, so I'm not sure how important it's to not break things.

### CODE REVIEW

File: Cartfile.private [hunk 1]
```diff
@@ -1,3 +1,4 @@
 github "Carthage/Commandant" ~> 0.15.0
+github "IBM-Swift/BlueSocket" ~> 1.0.0
```
Comment (marcelofabri) [inline Cartfile.private]: new dependency: are we ok with it?

File: Makefile [hunk 2]
```diff
@@ -119,7 +119,7 @@ archive:
 release: package archive portable_zip
 
 docker_test:
-	docker run -v `pwd`:`pwd` -w `pwd` --name swiftlint --rm norionomura/swift:42 swift test --parallel
+	docker run -v `pwd`:`pwd` -w `pwd` --name swiftlint --rm norionomura/swift:421 swift test --parallel
```
Comment (marcelofabri) [inline Makefile]: there's an issue on Linux when compiling with 4.2 which is fixed on 4.2.1 (should we highlight this bump somewhere)?

File: Source/SwiftLintFramework/Models/Linter.swift [hunk 3]
```diff
@@ -7,16 +7,83 @@ private struct LintResult {
     let deprecatedToValidIDPairs: [(String, String)]
 }
 
-private extension Rule {
-    static func superfluousDisableCommandViolations(regions: [Region],
-                                                    superfluousDisableCommandRule: SuperfluousDisableCommandRule?,
-                                                    allViolations: [StyleViolation]) -> [StyleViolation] {
+private enum LintableBox {
+    case rule(Rule)
+    case remoteRule(RemoteRule)
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/Linter.swift]: I'm not a big fan of having 2 totally different models, but it seemed a better idea than having something like `CustomRules` which gives us surprising results sometimes.

File: Source/SwiftLintFramework/Models/RemoteRule.swift [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                                        path: file.path,
+                                        contents: Lazy(file.contents),
+                                        configuration: configuration)
+        return validate(payload: payload, file: file)
+    }
+
+    private func validate(payload: RemoteRulePayload, file: File) -> [StyleViolation] {
+        do {
+            let socket = try Socket.create(family: .unix, type: .stream, proto: .unix)
+            try socket.connect(to: "/tmp/\(ruleDescription.identifier).socket")
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/RemoteRule.swift]: are we okay with hardcoding the path like this?

File: Source/SwiftLintFramework/Models/RemoteRulePayload.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        if path == nil && contents == nil {
+            return nil
+        }
+
+        self.init(structure: Lazy(structure), syntaxMap: Lazy(syntaxMap),
+                  path: path, contents: Lazy(contents), configuration: configuration)
+    }
+}
+
+private func convertingIntsToInt64(value: Any) -> Any {
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Models/RemoteRulePayload.swift]: not a big fan, but couldn't think about anything better

File: SwiftLint.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -2195,7 +2247,7 @@
 			isa = XCBuildConfiguration;
 			baseConfigurationReference = D0D1212819E878CC005E4BAA /* Release.xcconfig */;
 			buildSettings = {
-				MACOSX_DEPLOYMENT_TARGET = 10.10;
+				MACOSX_DEPLOYMENT_TARGET = 10.11;
```
Comment (marcelofabri) [inline SwiftLint.xcodeproj/project.pbxproj]: had to bump this because of `BlueSocket` - hopefully, it's ok since 10.11 is already pretty old by now

**Comment by therealbnut** [comment]
This looks fantastic, an epic undertaking. It should be really powerful and useful when it’s done. 💯

While we wait there’s a few neglected PRs that improve custom rules, which should allow people to start using them properly. IMO more custom rules will seed a more vibrant plugin community. The oldest open PR on this repo is such a PR, it was blocked 2 years ago with the only real comment saying to wait for plugins. I’ve made one more recently: 
https://github.com/realm/SwiftLint/pull/2586

I don’t think it’s productive to block custom rule improvements until plugins are ready, even if plugins will be ready in as little as a month. Custom rule improvements allow people to explore the possibilities plugins will allow.

Awesome work, I look forward to this landing! Please say if there’s any plugin tasks the community can help with. This weekend if I have time I might experiment with moving the existing rules into a separate module, so we can determine what APIs need to be public for plugins.

**Comment by marcelofabri** [comment]
> I don’t think it’s productive to block custom rule improvements until plugins are read

I don't think that's why these PRs weren't reviewed. It's probably just that no maintainer was able to dedicate the time to review them, as all maintainers work on SwiftLint on their free time.

> Awesome work, I look forward to this landing! Please say if there are any plugin tasks the community can help with. This weekend if I have time I might experiment with moving the existing rules into a separate module, so we can determine what APIs need to be public for plugins.

Thanks! I think that'd be an interesting exercise. 

Another update on plugins: I've talked with @jpsim a few weeks ago and we decided that we should explore having plugins as Swift files in a `.swiftlint` folder and then SwiftLint would compile them as one (or multiple - TBD) dynamic library. This has some benefits over my initial suggestion:

- Plugins can run in process and much faster
- Less complexity (we don't need to change much code after a `Rule` is registered from a plugin, no sockets, etc)
- Better experience for users

However, I wasn't able to get a proof of concept working. I've tried using SPM, but run into https://bugs.swift.org/browse/SR-6091. The problem is:

1. Plugin will need to depend on SwiftLintFramework dynamically 
2. We link SwiftLintFramework statically on the `swiftlint` binary
3. We'll open the plugin dylib on runtime with `dlopen`

Help with getting the right incantation would be super valuable! We don't necessarily need to use SPM,  but I thought it would be helpful for `(1)`.

**Pull request closed without merging**